autoload/denops/_internal/plugin.vim

```diff
@@ -21,31 +21,38 @@ function! denops#_internal#plugin#load(name, script) abort
   const l:script = denops#_internal#path#norm(a:script)
   const l:args = [a:name, l:script]
   let l:plugin = denops#_internal#plugin#get(a:name)
-  let l:plugin.state = s:STATE_LOADING
+  if l:plugin.state ==# s:STATE_RESERVED || l:plugin.state ==# s:STATE_FAILED
+    let l:plugin.state = s:STATE_LOADING
+  endif
   let l:plugin.script = l:script
-  let s:plugins[a:name] = l:plugin
   call denops#_internal#echo#debug(printf('load plugin: %s', l:args))
   call denops#_internal#server#chan#notify('invoke', ['load', l:args])
 endfunction
 
 function! denops#_internal#plugin#unload(name) abort
   let l:args = [a:name]
   let l:plugin = denops#_internal#plugin#get(a:name)
-  let l:plugin.state = s:STATE_UNLOADING
+  if l:plugin.state ==# s:STATE_LOADED
+    let l:plugin.state = s:STATE_UNLOADING
+  endif
   call denops#_internal#echo#debug(printf('unload plugin: %s', l:args))
   call denops#_internal#server#chan#notify('invoke', ['unload', l:args])
 endfunction
 
 function! denops#_internal#plugin#reload(name) abort
   const l:args = [a:name]
   let l:plugin = denops#_internal#plugin#get(a:name)
-  let l:plugin.state = s:STATE_LOADING
+  if l:plugin.state ==# s:STATE_LOADED
+    let l:plugin.state = s:STATE_UNLOADING
+  endif
   call denops#_internal#echo#debug(printf('reload plugin: %s', l:args))
   call denops#_internal#server#chan#notify('invoke', ['reload', l:args])
 endfunction
 
 function! s:DenopsSystemPluginPre() abort
   const l:name = matchstr(expand('<amatch>'), 'DenopsSystemPluginPre:\zs.*')
+  let l:plugin = denops#_internal#plugin#get(l:name)
+  let l:plugin.state = s:STATE_LOADING
```

Comment:
Why do we need to overwrite the `state` here?

Reply:
The "reload" operation causes a state transition within Deno. We need to receive a notification from Deno and update the state in Vim.

Reply:
So it seems the current state flow is like

```mermaid
stateDiagram-v2
reserved --> loading: denops#_internal#plugin#load
loading --> loading: DenopsSystemPluginPre
loading --> loaded: DenopsSystemPluginPost
loading --> failed: DenopsSystemPluginFail
failed --> loading: denops#_internal#plugin#load
loaded --> unloading: denops#_internal#plugin#unload
loaded --> unloading: denops#_internal#plugin#reload
unloading --> unloading: DenopsSystemPluginUnloadPre
unloading --> reserved: DenopsSystemPluginUnloadPost
unloading --> reserved: DenopsSystemPluginUnloadFail
```

It's a bit complicated because the state is changed by function or autocmd.

Do you think we can omit state change in `denops#_internal#plugin#XXXXX()` so that state only changes with `User` autocmd?

Reply:
I think it's a good idea to do all state management on the Deno side.

Reply:
There are use cases where to call `reload()` and immediately call `wait()`, in which case the `reload()` will not work properly unless update the state in the Vim API.